### ISSUE TITLE
Finally fix (server side) radar image processing

### ISSUE THREAD

**Issue by airmanopus** [airmanopus/weatherCheck#205]
*sigh* I've been putting this off because it worked; but the borrowed perl script I'm using to remove clutter and noise from radar images is ancient and now it's not doing the filtering. (All that actually happens is that the colors from the radar image that are usually noise are made transparent so they're "invisible".) 

Lots going on here.

Previously: cron job ran that downloaded each of the raw radar images, and ran a perl script that used an imagemagick library to remove several of the colors that are most generally just "noise" rather than "probably precip". The perl was ~~gettting~~ outdated, and besides the process was really, really slow. So the displayed radar was often old enough to be several revolutions of the actual radar old. Ugh. (Trying to run multiple convert processes in parallel on my webhost quite reliably tripped the runaway process filters as well.)

Currently: radar display is borked. The cron job was taking so long it was overlapping with the next kick, so as time went on there were a crapload of processes that just went to sleep. Issue here is that wget was making ipv6 requests, and these were timing out, but this needs to be overall fixed anyway; so fixing that part didn't really solve the overall issue. Processed radar images aren't having the noise filtered out properly anyway (see 'Previously' above).

The 'new' way:
- New script that downloads the raw data images, but it downloads them two at a time in parallel. It will still run as a cron job, but since it takes less than a minute to run (as far as I've tested) it can run fairly often. The fastest that a raw image comes from the radar is six minutes, plus whatever time it takes to get the image to NWS's web server, so we don't need to request it any more often than that. wget will only download an image if it's newer than the one we have, and we'll only copy a downloaded image into the .../images/radar/ dir if it's a newer image (somewhat arbitrarily, this means setting the cron interval to 8-10 minutes so we're not requesting new images *too* often).
- Processing (removing the noise/colors we don't want from an image) is done with a new PHP script that does exactly the same thing the perl script did, but seems to do it a ton faster. The wrinkle here is that the PHP outputs (and serves) an image that is the processed radar image; so we (meaning _I_) need to modify all of the kml files since they're looking for static images. This Should Work(tm). See https://github.com/airmanopus/weatherCheck/commit/a71a11a77de7a74df918884aeaf71df812177807. 
- More on processing: the original idea behind batch processing the raw images and then just displaying the processed ones was _supposed_ to mean that the client would always be getting a very current radar image; images were ending up being up to 30 minutes old, which was just borked. The client side still has to ask the server for the radar image, so that isn't changing, although it does take time for the PHP to make the new image every time it's requested. Even with that overhead, client is still getting a much much more current radar image than processing them as a batch; damn near real time.
- Even more on processing: this also makes possible a solution for #195, where we want to be able to control how much we're filtering in winter when really light snow (which can cause a lot more problems than really light rain) needs to show up on the radar.  Now we can add that kind of logic to weatherCheck/cgi-bin/radarFilterTest.php.

Remaining todo:
- [x] Change all of the kml files to refer to https://www.airmanopus.net/cgi-bin/radarFilterTest.php/?radarSite=<three letter code> instead of the static image in /images/processed and make sure the images are loading (occasional 500 error notwithstanding, see later notes)
- [x] Try to find a way to log (Sentry?) images loaded to the map via kml (see previous item) to track how often we're tripping procwatch. Failing that, hack something up that parses error_log and gets details (size of image especially) to track the errors.
- [x] Try to find a way to load transparent.gif and display that instead of the broken image image that results from above (see weatherCheck/images/radar/transparent.gif)

**Comment by airmanopus** [comment]
see #162 for more background on all of this radar stuff

**Comment by airmanopus** [comment]
- [ ] Make sure that when the client asks for a processed radar image via the kml file, that we're actually getting the image we should be; test case is FFC in GA's kml
- [ ] Make sure that when the client is asking for adjacent states' processed radar images, that's actually happening; test case is again GA (which has a bunch of states adjacent to it, and therefore a bunch of radar images to display)

**Comment by airmanopus** [comment]
Seeing this error when loading radar images via the new php:
Premature end of script headers: radarFilterTest.php, referer: https://www.airmanopus.net/weathercheck/

This is actually a _server_ issue; Dreamhost has a procwatch daemon running that's killing processes (like the one referenced above) for using too much CPU or RAM (see https://help.dreamhost.com/hc/en-us/articles/216540488-Why-did-procwatch-kill-processes-on-my-Shared-server-), and it makes perfect sense that shared hosting works this way. 

One easy answer is upgrade to a virtual private server that doesn't have this limitation, which is along the lines of "throw hardware at the problem". Were this project a product I wouldn't think twice about going that route, but it's an interesting challenge to see if I can keep the radar filtering process from tripping procwatch as well as a learning opportunity to try to record when its happening and log the occurrences.
So _for now_ (v0.7.0) I'm going to keep working with that.

**Comment by airmanopus** [comment]
This is what it looks like when client requests an image, and we trip procwatch.

[Thu Aug 31 15:07:32 2017] [error] [client n.n.n.n] Premature end of script headers: radarFilterTest.php, referer: https://www.airmanopus.net/weathercheck/

After messing with this for a while, I'm more convinced that I've reached the limit of what I can do with the resources I have running this on shared hosting (which is totally *not* Dreamhost's fault). 
- There's not much to optimize out of the filter process, because there's really not much process
- Adding more to what radarFilter does by adding checks for actual weather conditions, etc. (see #195) is going to also increase the resources that each client request for a radar image needs
 
Which doesn't change any of the _coding_ todos for this issue, so I'm going to finish those and close when that's done, but the actual app will be in a borked state until I upgrade to a VPS.

**Comment by airmanopus** [comment]
Just so _I_ remember this; potentially storing the kml data on the client (see #138) won't/shouldn't be affected by changing the kml to point to the new filtering stuff now, since the client will have to load all of the kml files at least once anyway (and since that's not implemented none of them will need to be refreshed).

**Comment by airmanopus** [comment]
Letting the cron job that handles downloading radar images from NWS run overnight to see if anything borks. Entirely possible that in changing the image links in the kml files I missed/borked one, so check that after sleep; checking this *must* be done from the app view (look at the map and see if there are any borken image indicators).

It is a longstanding thought that the kml files could be generated, and their information stored in a db or other structure and retrieved dynamically, rather than loading individual kml files (which are just text files). In terms of getting radar images onto the screen, it's likely a wash but it would make updating them less maddeningly typing boring.

The dynamically generated (processed) radar images are not appearing/being loaded on iOS. Might end up as a separate new issue.

**Comment by airmanopus** [comment]
Yup, too many requests being made by Safari (let's just say iOS), but that's quickly becoming something that's going to be a separate kind of solution (and therefore new issue).
